pkg/coscheduling/coscheduling.go

```diff
@@ -652,89 +670,158 @@ func (cs *Coscheduling) preemptionTag(pod *v1.Pod) {
+	for _, node := range nodes {
+		if cs.fitSelector(group.selector, node.Node()) {
+			taint, err := node.Taints()
+			if err != nil {
+				continue
+			}
+			if cs.doesTolerate(group.tolerations, taint) && cs.getUsedSlots(node, true) > 0{
+				selectedNodes[node.Node().Name] = node
+			}
 		}
-		return *podsList[i].Spec.Priority < *podsList[j].Spec.Priority
-	})
+	}
 
-	freed := 0
-	lastPg := ""
-	preemptionCandidates := map[int]*v1.Pod{}
-	for _, p := range podsList {
-		if *p.Spec.Priority >= group.priority {
-			break
-		} else if lastPg != "" && lastPg != p.Labels[PodGroupName] {
-			break
-		} else if _, ok := p.Labels["determined-preemption"]; ok {
+	needed := group.minAvailable - available
+
+	if len(selectedNodes) < needed {
+		klog.V(3).Infof("No preemption occurred. There are not enough nodes in the node pool.")
+		return false, 0
+	}
+
+	// for each node, find the highest priority pod and use that for preemption metric
+	// priorities contains the highest priority pod for each node
+	priorities := map[string]int{}
+	timestamps := map[string]time.Time{}
+	for _, node := range nodes {
+		maximum := 0
+		ts := time.Now()
+		if cs.getUsedSlots(node, true) == 0 {
 			continue
-		} else {
-			if cs.compareSelectors(group.selector, p.Spec.NodeSelector) {
-				taints := cs.getTaints(p.Spec.NodeName)
-				if taints != nil && cs.doesTolerate(p.Spec.Tolerations, taints) {
-					preemptionCandidates[freed] = p
-					freed += 1
-				}
+		}
+		for _, pod := range node.Pods() {
+			pgInfo, _ := cs.getOrCreatePodGroupInfo(pod, time.Now())
+			if _, ok := pod.Labels["determined-system"]; ok {
+				maximum = SystemPriority
+			} else if _, ok := pod.Labels["determined-cmd"]; ok {
+				maximum = SystemPriority
+			} else if _, ok := pod.Labels["determined"]; !ok { //only count determined pods for preemption
+				continue
+			}
+			if int(pgInfo.priority) > maximum {
+				maximum = int(*pod.Spec.Priority)
+				ts = pgInfo.timestamp
+			} else if int(pgInfo.priority) == maximum && pgInfo.timestamp.Before(ts) {
+				ts = pgInfo.timestamp
 			}
 		}
-
-		if freed+available >= pgMinAvailable && lastPg == "" {
-			lastPg = p.Labels[PodGroupName]
+		priorities[node.Node().Name] = maximum
+		timestamps[node.Node().Name] = ts
+	}
+
+	// the lowest priority and most recent node gets preempted first
+	var nodesToPreempt []string
+	for needed > 0 {
+		minKey := ""
+		minimum := math.MaxInt32
+		ts := time.Now()
+		for k, v := range priorities {
+			if v < minimum {
+				minimum = v
+				minKey = k
+				ts, _ = timestamps[k]
+			} else if v == minimum && timestamps[k].After(ts) {
+				minKey = k
+				ts, _ = timestamps[k]
+			}
+		}
+		if minimum >= int(group.priority) {
+			break
 		}
+		nodesToPreempt = append(nodesToPreempt, minKey)
+		delete(priorities, minKey)
+		needed -= 1
 	}
 
-	if freed+available < pgMinAvailable {
+	if needed > 0 {
 		klog.V(3).Infof("No preemption occurred. Not enough nodes are able to be freed")
```

Comment:
It's possible SOME preemption occurred, just not enough, right?

Reply:
At this point, no preemption has occurred. The code finds all preemptible nodes first, then decides if it's enough. If enough, it'll do the preemption. The actual act of preempting pods starts at L765.

If there isn't enough preemption, we don't preempt anything